paho/client.go

```diff
@@ -521,13 +533,52 @@ func (c *Client) incoming() {
 				}()
 				return
 			case packets.PINGRESP:
-				c.debug.Println("received PINGRESP")
-				c.PingHandler.PingResp()
+				go func() {
+					c.debug.Println("received PINGRESP")
+					c.cPingresp <- recv.Content.(*packets.Pingresp)
+				}()
 			}
 		}
 	}
 }
 
+func (c *Client) networkLoop () {
+	cIncoming := make(chan *packets.ControlPacket)
+	readNext := func() {
+		recv, err := packets.ReadPacket(c.Conn)
+		if err != nil {
+			c.debug.Printf("incoming loop error: %v",err)
+			go c.error(err)
+			return
+		} else {
+			c.debug.Println(fmt.Sprintf("Received: %v", recv.PacketType()))
+			cIncoming <- recv
+		}
+	}
+
+	go readNext()
+	for {
+		select {
+		case <-c.stop:
+			c.debug.Println("Loop stop")
+			return
+		case p := <- c.cOutgoingPackets:
+			_, err := p.WriteTo(c.Conn)
```

Comment:
Should writing to the connection be in a different goroutine from the network loop?

Reply:
I went back and forth on this  I ultimately decided that they sort-of belonged together, but I could be convinced to separate them again.